File: api/waitlist/window.py
from flask import Blueprint, g, request
import pydantic
from .webutil import ViewReturn
from . import auth
from .data import esi

bp = Blueprint("window", __name__)


class OpenWindowRequest(pydantic.BaseModel):
    target_id: int


@bp.route("/api/open_window", methods=["POST"])
@auth.login_required
@auth.select_character()
def open_window() -> ViewReturn:
    req = OpenWindowRequest.parse_obj(request.json)
    target_id = req.target_id

    esi.post("/v1/ui/openwindow/information/?target_id=%d" % target_id, g.character_id)
    return "OK"

File: backend/src/routes/window.rs
use rocket::serde::json::Json;
use serde::{Deserialize, Serialize};

use crate::{
    app::Application,
    core::{
        auth::{authorize_character, AuthenticatedAccount},
        esi::ESIScope,
    },
    util::madness::Madness,
};

#[derive(Deserialize, Debug)]
struct OpenWindowRequest {
    character_id: i64,
    target_id: i64,
}

#[derive(Deserialize, Serialize, Debug)]
struct Empty {}

#[post("/api/open_window", data = "<input>")]
async fn open_window(
    account: AuthenticatedAccount,
    app: &rocket::State<Application>,
    input: Json<OpenWindowRequest>,
) -> Result<(), Madness> {
    authorize_character(app.get_db(), &account, input.character_id, None).await?;

    app.esi_client
        .post(
            &format!(
                "/v1/ui/openwindow/information/?target_id={}",
                input.target_id
            ),
            &Empty {},
            input.character_id,
            ESIScope::UI_OpenWindow_v1,
        )
        .await?;

    Ok(())
}

pub fn routes() -> Vec<rocket::Route> {
    routes![open_window]
}
